### PR TITLE
API-1835: use discovery for the discovery endpoint

### PR DESCRIPTION
**Pull request by deads2k**
Updates the manifestclient to use aggregated discovery from https://github.com/openshift/oc/pull/1901 to provide discovery and provide ListKind for list requests.  This lets us have informers sync when we don't have any data for a type.

There's also a compromise in here to have namespaced lists where the namespace isn't in must-gather return empty list. While not true to the kube-apiserver, this allows all informers to sync.

### CODE REVIEW

File: pkg/manifestclient/group_resource_discovery.go [hunk 1]
```diff
@@ -4,14 +4,58 @@ import (
...
+	default:
+		// TODO can probably do better
+		return mrt.getLegacyGroupResourceDiscovery(requestInfo)
+	}
+}
+
+func (mrt *manifestRoundTripper) getAggregatedDiscoveryForURL(filename, url string) ([]byte, error) {
+	discoveryBytes, err := fs.ReadFile(mrt.sourceFS, filename)
+	if os.IsNotExist(err) {
+		discoveryBytes, err = fs.ReadFile(defaultDiscovery, filepath.Join("default-discovery", filename))
```
Comment (p0lyn0mial) [inline pkg/manifestclient/group_resource_discovery.go:40]: won't this get us into trouble? The `applycfg` cmd will provide the input directory, which could contain data from the current cluster discovery.

File: pkg/manifestclient/read_roundtripper.go [hunk 2]
```diff
@@ -128,10 +140,94 @@ func isServerGroupResourceDiscovery(path string) bool {
...
 	if len(parts) != 4 {
 		return false
 	}
 	return parts[0] == "" && parts[1] == "apis"
 }
+
+//go:embed default-discovery
+var defaultDiscovery embed.FS
+
+func (mrt *manifestRoundTripper) getKindForResource(gvr schema.GroupVersionResource) (kindData, error) {
```
Comment (p0lyn0mial) [inline pkg/manifestclient/read_roundtripper.go:157]: `kindData` leaks the `err` filed which is not used outside of this function, can we get that fixed?

File: pkg/manifestclient/read_roundtripper.go [hunk 3]
```diff
@@ -128,10 +140,94 @@ func isServerGroupResourceDiscovery(path string) bool {
...
+				if len(resourceInfo.ResponseKind.Group) > 0 {
+					kindForGVR.kind.Group = resourceInfo.ResponseKind.Group
+				}
+				if len(resourceInfo.ResponseKind.Version) > 0 {
+					kindForGVR.kind.Version = resourceInfo.ResponseKind.Version
+				}
+				kindForGVR.listKind = schema.GroupVersionKind{
+					Group:   kindForGVR.kind.Group,
+					Version: kindForGVR.kind.Version,
+					Kind:    resourceInfo.ResponseKind.Kind + "List",
```
Comment (p0lyn0mial) [inline pkg/manifestclient/read_roundtripper.go:222]: I think this won't work for CRDs.

File: pkg/manifestclient/group_resource_discovery.go (hunk 1 above)
Comment (deads2k) [reply pkg/manifestclient/group_resource_discovery.go:40]: > won't this get us into trouble? The `applycfg` cmd will provide the input directory, which could contain data from the current cluster discovery.

The logic (I think) is, "prefer the must-gather discovery, if must-gather discovery doesn't exist, use a default".  So it's discovery of last resort.  It let's us work with older must-gathers and incomplete ones.

File: pkg/manifestclient/read_roundtripper.go (hunk 2 above)
Comment (deads2k) [reply pkg/manifestclient/read_roundtripper.go:157]: > `kindData` leaks the `err` filed which is not used outside of this function, can we get that fixed?

How would you cache the error with out storing it in kindData?  We hit this path thousands of times.

File: pkg/manifestclient/read_roundtripper.go (hunk 3 above)
Comment (deads2k) [reply pkg/manifestclient/read_roundtripper.go:222]: > I think this won't work for CRDs.

Technically not correct.  Practically the best option we have.  Turns out we don't include ListKind in discovery.

**Comment by deads2k** [comment]
negotiated via slack. We think any misses will be caught during integration testing.  And it would have been nice if new-discovery had included ListKind :(

**Pull request merged**